dashboard/app/jobs/concerns/active_job_metrics.rb

```diff
@@ -0,0 +1,104 @@
+require 'cdo/aws/metrics'
+require 'cdo/honeybadger'
+
+module ActiveJobMetrics
+  extend ActiveSupport::Concern
+
+  METRICS_NAMESPACE = 'code-dot-org/ActiveJob'.freeze
+
+  included do
+    # Parent class callbacks are called in addition to any callbacks defined in the job subclass
+    # Callback functions are executed in the following order:
+    #   before_enqueue and around_enqueue(before) in any order
+    #   Job is added to database or other queue store
+    #   after_enqueue and around_enqueue(after) in any order
+    #   before_perform and around_perform(before) in any order
+    #   job is performed
+    #   after_perform and around_perform(after) in any order
+    # When both the Parent and Child class define the same callback, they may be called in any order.
+    # Also note that jobs executed via `.perform_now` will not trigger the `enqueue` callbacks.
+    # https://guides.rubyonrails.org/v6.0/active_job_basics.html#available-callbacks
+    after_enqueue :report_job_count
+    before_perform :report_wait_time
+    after_perform :report_performance
+  end
+
+  protected def common_dimensions
+    [
+      {name: 'Environment', value: CDO.rack_env},
+      {name: 'JobName', value: self.class.name},
+      {name: 'JobId', value: job_id},
+      {name: 'QueueName', value: queue_name},
+    ]
+  end
+
+  protected def report_job_count
+    Cdo::Metrics.push(
+      METRICS_NAMESPACE, [
+        {
+          # Same metric as "bin/cron/report_activejob_metrics"
+          metric_name: 'JobCount',
+          value: Delayed::Job.count,
+          unit: 'Count',
+          timestamp: Time.now,
+          dimensions: [
+            {name: 'Environment', value: CDO.rack_env},
+          ],
+        },
+      ]
+    )
+  rescue => exception
+    Honeybadger.notify(exception, error_message: 'Error reporting ActiveJob metrics')
+  end
```

Comment:
Today, with EvaluateRubricJob as the only job hitting the queue, we can assume that a non-zero `JobCount` is a possible indicator of a stuck queue. Now that we're adding a job that will sit in the queue for 24 hours, we should revisit what metrics are actually useful.

This metric is currently only reporting the environment metric - should we update it to use `common_dimensions` like the others?

Reply:
Depends on what you need this metric for. If you want to identify long-running jobs, we could count jobs using `Delayed::Job.where("locked_at IS NOT NULL AND locked_at < ?", 5.minutes.ago)` instead.

Reply:
Depends on what you need this metric for. If you want to identify long-running jobs, we could count them like:
```ruby
Delayed::Job.where("locked_at IS NOT NULL AND locked_at < ?", 5.minutes.ago)
```

Reply:
5 minutes seems like the right amount of time considering what we would expect the AI job timing to be.

Noticing that there are often jobs that take longer than, say, 20 minutes is valuable. Since that only will be recorded if that job completes, and we would want to know if there is a job that is stuck, we would need one of these actions (on enqueue, probably) to also post up info about the current runtime of the current running task (aka queue/wait time) along with the queue length. If both numbers go up, the queue isn't stable.

Reply:
If I want to get an early warning that our queue workers aren't keeping up (either too slow, or stopped), then we could filter this down to just available jobs:

```sql
WHERE run_at < now AND locked_at IS NULL
```

We want that to be zero usually. We could also track a `WaitTimeOfOldestWaitingJob` metric (needs a better name).

Reply:
Lets just add the JobName dimension to this metric for now. We could consider adding an "AvailableJobCount" metric later if we think it would be useful.

```suggestion
  protected def report_job_count
    Cdo::Metrics.push(
      METRICS_NAMESPACE, [
        {
          # Same metric as "bin/cron/report_activejob_metrics"
          metric_name: 'JobCount',
          value: Delayed::Job.count,
          unit: 'Count',
          timestamp: Time.now,
          dimensions: common_dimensions,
        },
      ]
    )
  rescue => exception
    Honeybadger.notify(exception, error_message: 'Error reporting ActiveJob metrics')
  end
```